[WIP] Refactor voxel grid (#4829)
PointCloudLibrary/pcl: doc/tutorials/content/voxel_grid_structure.rst

.. _grid_filters:

Introduction
============

> This document is intended to explain our new ``VoxelGrid`` design and help you implementing your own grid based filters easily by reusing our implementations.

Reviewer: ```suggestion
This document is intended to explain our new ``VoxelGrid`` design and help you implement your own grid based filters easily by reusing our building blocks.
```

VoxelGrid structure
===================

We will explain the pieces composing ``VoxelGrid``, then elaborate their purpose and demonstrate how they can be reused to implement your custom grid filters.

> In a high level view, ``VoxelGrid`` is comprised from three important classes:

Reviewer: ```suggestion
``VoxelGrid`` is composed using three important classes:
```

> 1. ``Voxel``: define the information storing in a single voxel and its operations

Reviewer: ```suggestion
1. ``Voxel``: define the information stored in a single voxel and its operations
```

> 2. ``VoxelStruct``: define ``VoxelGrid`` filtering logic

Reviewer: ```suggestion
2. ``VoxelStruct``: define the filtering logic
```

3. ``VoxelFilter``: implement common set of API for voxel grid filters

> Together they form ``VoxelFilter<VoxelStruct<Voxel<PointT>,PointT>>`` which is is aliased as ``VoxelGrid<PointT>`` internally.

Reviewer: Add something like: Your custom Voxel Grid Filter would look like: ``VoxelFilter<MyCustomLogic<MyVoxel>>``

====================

1. Voxel
--------

``Voxel<PointT>``

A grid cell class which stores the basic information of a single voxel and can downsample the points in the voxel.

Member functions
^^^^^^^^^^^^^^^^

* ``void add(const PointT&)``: add a single point to the voxel
* ``PointT get()``: downsample the added points and return a single point representing the voxel 
* ``void clear()``: clear the stored voxel information
* ``size_t size()``: return the number of points added to the voxel

.. note::
  It can either downsample all the data fields or just the XYZ fields of points in the voxel, you can indicate that by passing a ``bool`` to the constructor.  
   
2. VoxelStruct
--------------

``VoxelStruct<Voxel<PointT>,PointT>``

It implements ``VoxelGrid`` filtering logic and accepts template argument ``Voxel<PointT>`` as the grid cell type.

Three general phases of grid based filters
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

1. Setting up the filter given necessary information
2. Grouping points to specific voxels based on the points' information
3. Selecting eligible voxels and downsampling each of them to a single point for pushing to the output point cloud

> They are implemented in ``VoxelStruct`` as ``VoxelGrid`` filtering logic and will be used by ``TransformFilter``.

Reviewer: Reword this. Slightly comfusing. What are you trying to get across?

.. note::
  ``TransformFilter`` is the base class of ``VoxelGrid`` and accepts template argument ``VoxelStruct`` as the filter logic. 
  It implements the general phases of grid based filters and has two outer loops to iterate through the points (phase 2) and non-empty voxels (phase 3).

  Please see ``TransformFilter`` for more information.
  
Member functions
^^^^^^^^^^^^^^^^

* ``size_t size()``
  
  * Return the number of the non-empty voxels
  
* ``begin()`` and ``end()``
  
  * Used for iterating through the voxels

>   * Return any type which can be incremented and compared, e.g., iterator, index

Reviewer: Can be non-member functions

* ``bool setUp(CartesianFilter<pcl::Filter, VoxelStruct>&)``
  
  * Initialize the member variables for filtering
  * Input point cloud and all settings of CartesianFilter can be accessed from the reference

* ``void addPointToGrid(PointT&)``

  * Add a single point to a specific voxel

* ``pcl::experimental::optional<PointT> filterGrid(...)``

  * Accept output from ``begin()`` and ``end()``
  * Decide whether the iterating voxel is eligible
  * Push a point representing the iterating voxel

3. VoxelFilter
--------------

``VoxelFilter<VoxelStruct<Voxel<PointT>,PointT>>``

It implements the API specific for ``VoxelGrid`` and accepts template argument ``VoxelStruct`` as the filtering logic.

Example: Create a custom grid filter
==========================

An example is worth a throusand words, let's implement a simple grid filters easily by reusing our classes!

1. Define a custom grid cell
----------------------------

This is a simple grid cell example which stores the number of points in the cell and can accumulate XYZ coordinates, 
and it can compute the centroid by averaging.

You should determine the information you need to store for your custom grid cell.

.. note::

>   The following member functions are not necessary if you are not using our ``VoxelStruct``.

Reviewer: ```suggestion
  The following member functions are not necessary if you are not using ``pcl::VoxelStruct``.
```

.. code-block:: cpp

>   template <typename PointT>

Reviewer: No template in example please. Let's keep it simple :)

  struct MyVoxel {
    // Accumulate points
    inline void add(const PointT& pt) {
      num_pt_++;
      centroid_ += pt.getArray4fMap();
    }

    // Export a point to the output
    inline PointT get() const {
      PointT pt;
      pt.getArray4fMap() = centroid_ / num_pt_;
      return pt;
    }

    // Empty the voxel
    inline void clear() {
      num_pt_ = 0;
      centroid_ = Eigen::Array4f::Zero();
    }

    // Return the number of added points
    inline std::size_t size() const { return num_pt_; }

  protected:
    // Enable Eigen SSE with Array4f instead Array3f 
    Eigen::Array4f centroid_ = Eigen::Array4f::Zero();
    std::size_t num_pt_ = 0;
  };

.. note::
  **Tips**:

>   If your application needs to filter some incoming point clouds repeatedly,

Reviewer: Is this tip relevant for `MyVoxel`? There might be a better place for this top

  reusing the same filter object instead of creating a new one is way faster.
  It is because cpp don't have to allocate the grid cells again and again.

2. Define a custom VoxelStruct
------------------------------

> This is a simple custom ``VoxelStruct`` withs a fixed grid size and fixed condition for selecting eligible voxels.

Reviewer: ```suggestion
This is a simple custom ``VoxelStruct`` with a fixed grid size and fixed condition for selecting eligible voxels.
```

We also implement a custom point hashing function ``myHash`` to select which grid cells we want to add our points to.
(you can also use our implementation ``hashPoint<PointT>(...)`` but for simplity we don't use that here)

.. note::
  The following member functions except our custom ``myHash`` are mandatory.

> .. code-block:: cpp

Reviewer: Would it be possible to instead add a CMakeLists.txt + source file and refer to the lines in that instead?

      it->second.add(pt);
    }

    // Decide if we want to export a point to the output given a grid cell
    inline optional<PointT> filterGrid(const iterator grid_it) {
      const auto& voxel = grid_it->second;
      if (voxel.size() >= min_points_per_voxel_)
        return voxel.get();
      else
        return boost::none;
    }

  protected:
    // A simple function that hash a point
    std::size_t myHash(const PointT& pt) {
      std::size_t hash = 0;
      const int grid_x = static_cast<int>(std::floor(pt.x / grid_size_));
      const int grid_y = static_cast<int>(std::floor(pt.y / grid_size_));
      const int grid_z = static_cast<int>(std::floor(pt.z / grid_size_));
      boost::hash_combine(hash, grid_x);
      boost::hash_combine(hash, grid_y);
      boost::hash_combine(hash, grid_z);
      return hash;
    }

    std::string filter_name_;
    Grid grid_;
    float grid_size_ = 5;
    std::size_t min_points_per_voxel_ = 5;
  };

3. Put everything together
--------------------------

With our custom ``Voxel`` and ``VoxelStruct``, we can pass them to ``TransformFilter`` and it will handle everything for you.

> Here we pass ``pcl::Filter`` for our non binary removal filter. For binary removal filters, you should pass ``pcl::FilterIndices`` .

Reviewer: What's binary removal?

.. code-block:: cpp

  template <typename PointT>

>   using MyVoxelGrid = TransformFilter<pcl::Filter, MyVoxelStruct<PointT>, PointT>;

Reviewer: Can we use VoxelFilter here instead of TransformFilter?
